docs/tutorial/native-code-and-electron.md

```diff
@@ -0,0 +1,375 @@
+
+This tutorial assumes you have Node.js and npm installed, as well as the basic tools necessary for compiling code on your platform (like Visual Studio on Windows, Xcode on macOS, or GCC/Clang on Linux). You can find detailed instructions in the [`node-gyp` readme](https://github.com/nodejs/node-gyp?tab=readme-ov-file).
+
+### Requirements: macOS
+
+To build native Node.js addons on macOS, you'll need the Xcode Command Line Tools. These provide the necessary compilers and build tools (namely, `clang`, `clang++`, and `make`). The following command will prompt you to install the Command Line Tools if they aren't already installed.
+
+```sh
+xcode-select --install
+```
+
+### Requirements: Windows
+
+The official Node.js installer offers the optional installation of "Tools for Native Modules", which installs everything required for the basic compilation of C++ modules - specifically, Python 3 and the "Visual Studio Desktop development with C++" workload. Alternatively, you can use `chocolatey`, `winget`, or the Windows Store.
+
+### Requirements: Linux
+
+* [A supported version of Python](https://devguide.python.org/versions/)
+* `make`
+* A proper C/C++ compiler toolchain, like [GCC](https://gcc.gnu.org)
+
+## 1) Creating a package
+
+First, create a new Node.js package that will contain your native addon:
+
+```sh
+mkdir my-native-addon
+cd my-native-addon
+npm init -y
+```
+
+This creates a basic `package.json` file. Next, we'll install the necessary dependencies:
+
+```sh
+npm install --save node-addon-api bindings
+```
+
+* `node-addon-api`: This is a C++ wrapper for the low-level Node.js API that makes it easier to build addons. It provides a C++ object-oriented API that's more convenient and safer to use than the raw C-style API.
+* `bindings`: A helper module that simplifies the process of loading your compiled native addon. It handles finding your compiled .node file automatically.
+
+Now, let's update our package.json to include the appropriate build scripts. We will explain what these specifically do further below.
+
+```json
+{
+  "name": "my-native-addon",
+  "version": "1.0.0",
+  "description": "A native addon for Electron",
+  "main": "js/index.js",
+  "scripts": {
+    "clean": "rm -rf build",
+    "build": "node-gyp configure && node-gyp build"
+  },
+  "dependencies": {
+    "bindings": "^1.5.0",
+    "node-addon-api": "^8.3.0"
+  }
+}
+```
+
+These scripts will:
+
+* `clean`: Remove the build directory, allowing for a fresh build
+* `build`: Run the standard node-gyp build process to compile your addon
+
+## 2) Setting up the build system
+
+Node.js addons use a build system called `node-gyp`, which is a cross-platform command-line tool written in Node.js. It compiles native addon modules for Node.js using platform-specific build tools behind the scenes:
```

Comment:
Should clarify here that this is specifically for C++ (since we mention Rust alternatives later on in this doc).

Reply:
It's not, you'll need all the same requirements even if you use other languages.

https://napi.rs/docs/deep-dive/native-module#node-gyp

Reply:
I'm out of my depth here, but from the [`napi-rs` README](https://github.com/napi-rs/napi-rs/blob/a25504a577e299101f4a541a2fd43fa2484f6166/crates/napi/README.md?plain=1#L51), the project doesn't seem to have a dependency on node-gyp. 🤔

> One nice feature is that this crate allows you to build add-ons purely with the Rust/JavaScript toolchain and without involving node-gyp.

Likewise, I _think_ `node-gyp` was only used in Neon for the Nan-based `neon-sys` crate (removed in the v1 milestone last year).

Reply:
Resolved in Slack: We're sticking with official Node tooling for now and will keep all of this around, we might add more dedicated tutorials just for Rust in the future.